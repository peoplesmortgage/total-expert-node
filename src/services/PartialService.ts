import { Service } from 'typedi';
import AuthenticatedService from './AuthenticatedService';
import { TotalExpertId, GetManyResponse, CreatedResponse } from '../types';

@Service()
class TEService<T, E> extends AuthenticatedService {
  constructor(urlSlug: string) {
    super(urlSlug);
  }

  protected async getToJSON(url: string): Promise<any> {
    const response = await this.auth.withAuthFetch(url);
    const data = await response.json();
    return data;
  }

  async create(providedData: E): Promise<CreatedResponse> {
    const options = { method: 'POST', body: JSON.stringify(providedData) };
    const response = await this.auth.withAuthFetch(this.urlSlug, options);
    const data = await response.json();
    return data;
  }

  async get(id: TotalExpertId): Promise<T> {
    const result = await this.getToJSON(`${this.urlSlug}/${id}`);
    return result;
  }

  async getMany(pageNumber: number = 0, pageSize: number = 10): Promise<GetManyResponse<T>> {
    const url = `${this.urlSlug}?page[number]=${pageNumber}&page[size]=${pageSize}`;
    return this.getToJSON(url);
  }
}

export default TEService;
