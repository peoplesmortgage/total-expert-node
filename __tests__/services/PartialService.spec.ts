import 'reflect-metadata';
import PartialService from '../../src/services/PartialService';
import mockAuthentication, { mockAPIReturnValue } from '../__fixtures__/mockAuthentication';

describe('PartialService', () => {
  const mockAuth = mockAuthentication();

  beforeEach(() => {
    jest.resetAllMocks();
  });

  it('stores its constructor argument as the urlSlug protected property', () => {
    expect(new PartialService<any, any>('/the-url-slug-value')).toMatchSnapshot();
  });

  ['create', 'get', 'getMany'].forEach((method) => {
    it(`exposes the ${method} method`, () => {
      const wrapper = new PartialService<any, any>('/some-value') as any;
      expect(typeof wrapper[method] as any).toEqual('function');
    });
  });

  ['get', 'getMany'].forEach((method) => {
    it(`exposes a ${method} method that returns the json response from an API request`, async () => {
      const mockResponseJSONValue = 'the mock response value';
      mockAPIReturnValue(mockResponseJSONValue);
      const wrapper = new PartialService('/') as any;
      const result = await wrapper[method]('test-id') as any;

      expect.assertions(2);
      expect(mockAuth.withAuthFetch).toHaveBeenCalledTimes(1);
      expect(result).toEqual(mockResponseJSONValue);
    });
  });

  it('will create a record with the arguments provided', async () => {
    mockAPIReturnValue({});
    await new PartialService('/the-resource-slug').create({ mock: 'test data' });

    const fetchArgs = mockAuth.withAuthFetch.mock.calls[0];
    expect(fetchArgs).toMatchSnapshot();
  });

  it('will update the page return parameters based off the arguments provided to getMany', async () => {
    mockAPIReturnValue({});
    await new PartialService('/').getMany(10, 100);
    const requestedUrl = mockAuth.withAuthFetch.mock.calls[0][0];
    const queryString = requestedUrl.split('?')[1];
    expect(queryString).toEqual('page[number]=10&page[size]=100');
  });

  it('will return the first 10 records of page 0 if no arguments are provided to getMany', async () => {
    mockAPIReturnValue({});
    await new PartialService('/').getMany();
    const requestedUrl = mockAuth.withAuthFetch.mock.calls[0][0];
    const queryString = requestedUrl.split('?')[1];
    expect(queryString).toEqual('page[number]=0&page[size]=10');
  });
});
