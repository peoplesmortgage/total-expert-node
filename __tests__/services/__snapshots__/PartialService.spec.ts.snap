// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PartialService stores its constructor argument as the urlSlug protected property 1`] = `
TEService {
  "auth": Object {
    "authenticate": [MockFunction],
    "setAccessToken": [MockFunction],
    "withAuthFetch": [MockFunction],
  },
  "urlSlug": "/the-url-slug-value",
}
`;

exports[`PartialService will create a record with the arguments provided 1`] = `
Array [
  "/the-resource-slug",
  Object {
    "body": "{\\"mock\\":\\"test data\\"}",
    "method": "POST",
  },
]
`;
